### PR TITLE
A few speed up.

### PR DESCRIPTION
**Pull request by nouiz**

### CODE REVIEW

File: theano/gpuarray/opt.py [hunk 1]
```diff
@@ -273,8 +277,12 @@ def apply(self, fgraph):
         # Building a new graph
         # Iterating through inputs of graph
         for i in fgraph.inputs:
-            if isinstance(i.type, tensor.TensorType):
-                mapping[i] = as_gpuarray_variable(i, None)  # TODO context
+            # Do not move *int* scalar to the GPU.
+            target = getattr(input.tag, 'target', None)
```
Comment (Sentient07) [inline theano/gpuarray/opt.py]: This condition was from op_lifter, right ?

File: theano/tensor/nnet/abstract_conv.py [hunk 2]
```diff
@@ -109,7 +109,16 @@ def get_conv_shape_1axis(image_shape, kernel_shape,
         pad = border_mode
         if pad < 0:
             raise ValueError("border_mode must be >= 0")
-    out_shp = (image_shape + 2 * pad - kernel_shape) // subsample + 1
+
+            # In case of symbolic shape, we want to build the smallest graph
```
Comment (Sentient07) [inline theano/tensor/nnet/abstract_conv.py:113]: I am able to understand what this code does, but how would it help in speeding up ?

File: theano/gpuarray/dnn.py [hunk 3]
```diff
@@ -957,9 +957,13 @@ def dnn_conv(img, kerns, border_mode='valid', subsample=(1, 1),
     desc = GpuDnnConvDesc(border_mode=border_mode, subsample=subsample,
                           conv_mode=conv_mode, precision=precision)(kerns.shape)
     desc_op = desc.owner.op
-    out_shp = GpuDnnConv.get_out_shape(img.shape, kerns.shape,
-                                       desc_op.border_mode,
-                                       desc_op.subsample)
+    # We can use Shape_i and bypass the infer_shape here as this is on
+    # the input of node and it will always be present.
+    ishape = [shape_i_op(i)(img) for i in range(img.ndim)]
```
Comment (Sentient07) [inline theano/gpuarray/dnn.py:962]: I don't understand this, could you please explain more on this ?

File: theano/tensor/nnet/abstract_conv.py (hunk 2 above)
Comment (nouiz) [inline theano/tensor/nnet/abstract_conv.py:113]: It speed things up as we will build a smaller graph. If pad is 0, it is useless to add 2*0 in the graph. If subsample is 1, there is no need to divide by 1.

Mostly, this apply a few elemwise optimization manually.

File: theano/gpuarray/opt.py [hunk 4]
```diff
@@ -273,8 +277,12 @@ def apply(self, fgraph):
         # Building a new graph
         # Iterating through inputs of graph
         for i in fgraph.inputs:
-            if isinstance(i.type, tensor.TensorType):
-                mapping[i] = as_gpuarray_variable(i, None)  # TODO context
+            # Do not move *int* scalar to the GPU.
+            target = getattr(i.tag, 'target', None)
```
Comment (nouiz) [inline theano/gpuarray/opt.py:281]: This was in InputToGpuOptimizer

File: theano/gpuarray/dnn.py (hunk 3 above)
Comment (nouiz) [inline theano/gpuarray/dnn.py:962]: This is a different way to get the shape.

a_var.shape[i] will create a small graph Subtensor(Shape(a_var), i). This create Shape_i(i)(a_var). So a smaller graph.

**Pull request merged**